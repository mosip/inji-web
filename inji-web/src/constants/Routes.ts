export const Pages = {
    ROOT: "/",
    USER: "user",
    HOME: "home",
    ISSUERS: "issuers",
    ISSUER_TEMPLATE: "issuers/:issuerId",
    CREDENTIALS: "credentials",
    PROFILE: "profile",
    FAQ: "faq",
    REDIRECT: "redirect",
    RESET_WALLET: "reset-wallet",
    PASSCODE: "passcode",
    AUTHORIZE: "authorize"
}

export const ROUTES = {
    ROOT: Pages.ROOT,
    PASSCODE: `/${Pages.USER}/${Pages.PASSCODE}`,
    USER_HOME: `/${Pages.USER}/${Pages.HOME}`,
    USER_ISSUERS: `/${Pages.USER}/${Pages.ISSUERS}`,
    USER_ISSUER: (id: string) => `/${Pages.USER}/${Pages.ISSUERS}/${id}`,
    ISSUER: (id: string) => `/${Pages.ISSUERS}/${id}`,
    CREDENTIALS: `/${Pages.USER}/${Pages.CREDENTIALS}`,
    PROFILE: `/${Pages.USER}.${Pages.PROFILE}`,
    FAQ: `/${Pages.FAQ}`,
    USER_FAQ: `/${Pages.USER}/${Pages.FAQ}`,
    USER_RESET_WALLET: `/${Pages.USER}/${Pages.RESET_WALLET}`,
};

