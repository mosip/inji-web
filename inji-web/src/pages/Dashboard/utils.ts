export const convertStringIntoPascalCase = (text: string | null) => {
    if (text != null) {
        return text
            .toLocaleLowerCase()
            .split(" ")
            .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
            .join(" ");
    }
};
