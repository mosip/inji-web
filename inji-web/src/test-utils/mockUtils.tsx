import React, { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter, Route, BrowserRouter as Router, Routes } from 'react-router-dom';
import { reduxStore } from '../redux/reduxStore';

// Mock for storage module
export const mockStorageModule = () => {
    jest.mock('../utils/storage.ts', () => ({
        storage: {
            getItem: jest.fn(),
            setItem: jest.fn(),
            SESSION_INFO: 'SESSION_INFO',
            SELECTED_LANGUAGE: 'selectedLanguage'
        }
    }));
};

// Mock for localStorage
export const mockLocalStorage = () => {
    let store: { [key: string]: string } = {};
  
    const localStorageMock = {
      getItem: jest.fn((key: string) => store[key] || null),
      setItem: jest.fn((key: string, value: string) => {
        store[key] = value;
      })
    };
  
    Object.defineProperty(window, 'localStorage', { 
      value: localStorageMock,
      writable: true
    });
  
    return localStorageMock;
  };

// API Mock
export const mockApi = {
    authorizationRedirectionUrl: 'http://mockurl.com'
};

// Crypto Mock
export const mockCrypto = {
    getRandomValues: (array: Uint32Array) => {
        for (let i = 0; i < array.length; i++) {
            array[i] = 4294967296;
        }
        return array;
    },
    subtle: {} as SubtleCrypto,
    randomUUID: () => '123e4567-e89b-12d3-a456-426614174000'
} as Crypto;

export const mockUseTranslation = () => {
    jest.mock('react-i18next', () => ({
        useTranslation: () => ({
            t: (key: string) => key,
        }),
    }));
};

export const mockUseNavigate = () => {
    const mockNavigate = jest.fn();
    jest.mock('react-router-dom', () => ({
        ...jest.requireActual('react-router-dom'),
        useNavigate: mockNavigate,
    }));
};

export const mockUseSelector = () => {
    jest.mock('react-redux', () => ({
        ...jest.requireActual('react-redux'),
        useSelector: jest.fn(),
    }));
};

export const setMockUseSelectorState = (state: any) => {
    const useSelectorMock = require('react-redux').useSelector;
    useSelectorMock.mockImplementation((selector: any) => selector(state));
};

export const mockUseGetObjectForCurrentLanguage = () => {
    jest.mock('../utils/i18n', () => ({
        getObjectForCurrentLanguage: jest.fn(),
    }));
};

export const wrapUnderRouter = (children: React.ReactNode) => {
    return <Router>{children}</Router>;
};

export const mockUseSearchCredentials = () => {
    jest.mock('../components/Credentials/SearchCredential', () => ({
        SearchCredential: () => <></>,
    }));
 };

export const mockUseDispatch = () => {
    jest.mock('react-redux', () => ({
        ...jest.requireActual('react-redux'),
        useDispatch: jest.fn(),
    }));
};
              
export const mockUseSpinningLoader = () => {
    jest.mock('../components/Common/SpinningLoader', () => ({
        SpinningLoader: () => <></>,
    }));
};

export const mockUseLanguageSelector = () => {
    jest.mock('../components/Common/LanguageSelector', () => ({
        LanguageSelector: () => <></>,
    }));
};

interface RenderWithProviderOptions extends Omit<RenderOptions, 'queries'> {}

export const renderWithProvider = (element: ReactElement, options?: RenderWithProviderOptions) => {
    return render(
        <Provider store={reduxStore}>
            <Router>
                {element}
            </Router>
        </Provider>,
        options
    );
};
    
export const mockUseParams = ()=>{
    jest.mock('react-router-dom', () => ({
        ...jest.requireActual('react-router-dom'),
        useParams: () => ({ issuerId: 'test-issuer-id' }),
      }));   
};

export const mockUseapiObject = () =>{
    jest.mock('../utils/api', () => ({
        api: {
          mimotoHost: 'http://mocked-api-host',
        },
      }));
};

export const mockUseFetch = () =>{
    jest.mock('../hooks/useFetch', () => {
        const RequestStatus = {
          LOADING: 'LOADING',
          DONE: 'DONE',
          ERROR: 'ERROR',
        };
        return {
          useFetch: () => jest.fn(),
          RequestStatus,
        };
      });
};

export const mockUseToast = () =>{
    jest.mock('react-toastify', () => ({
      toast: {
        error: jest.fn(),
      },
    }));
};

export const mockusemisc = ()=>{
    jest.mock('../utils/misc', () => ({
        downloadCredentialPDF: jest.fn(),
        getErrorObject: jest.fn(),
        getTokenRequestBody: jest.fn(),
      }));
};

export const mockusei18n = ()=>{
    jest.mock('react-i18next', () => ({
        useTranslation: () => ({
          t: (key: string) => key,
        }),
        initReactI18next: {
          type: '3rdParty',
          init: jest.fn(),
        },
      }));

};

export const mockWindowLocation = (url: string) => {
    const location = new URL(url) as unknown as Location;
    Object.defineProperty(window, 'location', {
      value: location,
      writable: true,
    });
  };

export const renderWithRouter = (Element: React.ReactElement, { route = '/' } = {}) => {
    window.history.pushState({}, 'Test page', route);
    return render(
      <BrowserRouter>
        <Provider store={reduxStore}>
          <Routes>
            <Route path="*" element={Element} />
          </Routes>
        </Provider>
      </BrowserRouter>
    );
  };

