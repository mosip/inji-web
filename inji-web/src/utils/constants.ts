export const KEYS = {
    USER: 'user',
    WALLET_ID: 'walletId',
};

// This constant exposes the pages used in the application. This will be used only by the Router component.
export const Pages = {
    ROOT: "/",
    USER: "user",
    HOME: "home",
    ISSUERS: "issuers",
    ISSUER_TEMPLATE: "issuers/:issuerId",
    CREDENTIALS: "credentials",
    PROFILE: "profile",
    FAQ: "faq",
    REDIRECT: "redirect",
    RESET_PASSCODE: "reset-passcode",
    PASSCODE: "passcode",
    AUTHORIZE: "authorize"
}

// This constant exposes the routes of the application which can be used across the component to avoid hardcoded navigation urls
export const ROUTES = {
    ROOT: Pages.ROOT,
    USER: `/${Pages.USER}`,
    PASSCODE: `/${Pages.USER}/${Pages.PASSCODE}`,
    USER_HOME: `/${Pages.USER}/${Pages.HOME}`,
    USER_ISSUERS: `/${Pages.USER}/${Pages.ISSUERS}`,
    ISSUERS: `/${Pages.ISSUERS}`,
    USER_ISSUER: (id: string) => `/${Pages.USER}/${Pages.ISSUERS}/${id}`,
    ISSUER: (id: string) => `/${Pages.ISSUERS}/${id}`,
    CREDENTIALS: `/${Pages.USER}/${Pages.CREDENTIALS}`,
    PROFILE: `/${Pages.USER}/${Pages.PROFILE}`,
    FAQ: `/${Pages.FAQ}`,
    USER_FAQ: `/${Pages.USER}/${Pages.FAQ}`,
    USER_RESET_PASSCODE: `/${Pages.USER}/${Pages.RESET_PASSCODE}`,
} as const;

export const HTTP_STATUS_CODES = {
    BAD_REQUEST: 400,
    UNAUTHORIZED: 401,
    FORBIDDEN: 403,
    NOT_FOUND: 404,
    INTERNAL_SERVER_ERROR: 500,
    SERVICE_UNAVAILABLE: 503,
}

export const PDF_WORKER_URL = "https://unpkg.com/pdfjs-dist@3.11.174/build/pdf.worker.min.js";
