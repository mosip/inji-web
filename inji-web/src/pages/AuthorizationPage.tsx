import React, {useEffect, useState} from "react";
import {SpinningLoader} from "../components/Common/SpinningLoader";
import {api} from "../utils/api";
import {LandingPageWrapper} from "../components/Common/LandingPageWrapper";
import {ErrorSheildIcon} from "../components/Common/ErrorSheildIcon";
import {useTranslation} from "react-i18next";

export const AuthorizationPage: React.FC = () => {
    const url = window.location.href;
    const currentQueryParams = new URLSearchParams(window.location.search);
    const [error, setError] = useState<String>(currentQueryParams.get("error")+"");
    const {t} = useTranslation("AuthorizationPage");

    useEffect( () => {
        if(url.indexOf("error") === -1) {
            window.location.href = api.mimotoHost + "/authorize" + window.location.search;
        }
    },[])

    if(url.indexOf("error") == -1){
        return <div><LandingPageWrapper icon={<SpinningLoader/>} title={""} subTitle={""} gotoHome={false}/></div>
    }
    return <div><LandingPageWrapper icon={<ErrorSheildIcon />} title={t(`error.code.${error}`)} subTitle={t(`error.message.${error}`)} gotoHome={true}/></div>
}


