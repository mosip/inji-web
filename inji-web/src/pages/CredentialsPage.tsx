import React, {useEffect, useState} from "react";
import {useParams} from "react-router-dom";
import {RequestStatus, useFetch} from "../hooks/useFetch";
import {NavBar} from "../components/Common/NavBar";
import {CredentialList} from "../components/Credentials/CredentialList";
import {useDispatch, useSelector} from "react-redux";
import {storeSelectedIssuer} from "../redux/reducers/issuersReducer";
import {storeCredentials, storeFilteredCredentials} from "../redux/reducers/credentialsReducer";
import {api} from "../utils/api";
import {useTranslation} from "react-i18next";
import {toast} from "react-toastify";

import {ApiRequest, DisplayArrayObject, IssuerObject} from "../types/data";
import {getObjectForCurrentLanguage} from "../utils/i18n";
import {RootState} from "../types/redux";
import {isObjectEmpty} from "../utils/misc";

export const CredentialsPage: React.FC = () => {

    const {state, fetchRequest} = useFetch();
    const params = useParams<CredentialParamProps>();
    const dispatch = useDispatch();
    const {t} = useTranslation("CredentialsPage");
    const language = useSelector((state: RootState) => state.common.language);
    let displayObject= {} as DisplayArrayObject;
    let [selectedIssuer, setSelectedIssuer] = useState({} as IssuerObject)
    if(!isObjectEmpty(selectedIssuer)){
        displayObject = getObjectForCurrentLanguage(selectedIssuer.display, language);
    }

    useEffect(() => {
        const fetchCall = async () => {
            let apiRequest: ApiRequest = api.fetchSpecificIssuer;
            let response = await fetchRequest(
                apiRequest.url(params.issuerId ?? ""),
                apiRequest.methodType,
                apiRequest.headers()
            );
            dispatch(storeSelectedIssuer(response?.response));
            setSelectedIssuer(response?.response);

            apiRequest = api.fetchCredentialTypes2;
            response = await fetchRequest(
                apiRequest.url(params.issuerId ?? ""),
                apiRequest.methodType,
                apiRequest.headers()
            );
            // apiRequest = api.fetchCredentialTypes;
            // response = await fetchRequest(
            //     apiRequest.url(response?.response[".well-known"]),
            //     apiRequest.methodType,
            //     apiRequest.headers()
            // );
            dispatch(storeFilteredCredentials(response?.response?.supportedCredentials));
            dispatch(storeCredentials(response?.response?.supportedCredentials));
        }
        fetchCall();
    }, [])

    if (state === RequestStatus.ERROR) {
        toast.error(t("errorContent"));
    }


    return <React.Fragment>
        <div className="bg-iw-background min-h-screen"
             data-testid="Credentials-Page-Container">
            <NavBar title={displayObject?.name} search={true} fetchRequest={fetchRequest} link={"/"}/>
            <div data-testid="Credential-List-Container" className="container mx-auto mt-8 px-10 sm:px-0">
                <CredentialList state={state}/>
            </div>
        </div>
    </React.Fragment>
}

