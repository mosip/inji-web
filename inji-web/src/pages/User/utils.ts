import { User } from "../../components/Dashboard/types";
import { KEYS } from "../../utils/constants";
import {ROUTES} from "../../constants/Routes";

export const convertStringIntoPascalCase = (text: string | undefined) => {
    return (
        text?.toLocaleLowerCase()
            .split(' ')
            .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
            .join(' ')
    );
};

export const navigateToUserHome = (navigate: any) =>
    navigate(ROUTES.USER_HOME);

export const validateWalletUnlockStatus = (
    cachedWalletId: string | null,
    storageWalletId: string | null,
    navigate: (path: string) => void,
    user: User
) => {
    if (cachedWalletId && (cachedWalletId === storageWalletId)) {
        console.info('Wallet is unlocked!');
    } else {
        console.warn(
            'Wallet exists but is locked, redirecting to `/pin` to unlock the wallet.'
        );
        if(user) {
            navigate('/pin');
            localStorage.removeItem(KEYS.WALLET_ID);
        }
    }
};