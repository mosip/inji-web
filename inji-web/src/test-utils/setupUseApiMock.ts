import {RequestStatus} from "../utils/constants";

export const mockUseApi = {
    fetchData: jest.fn(),
    state: RequestStatus.LOADING,
    data: null as any,
    error: null as any,
    status: null,
    ok: () => false
};

type MockApiResponseOptions = {
    data?: object;
    status?: number | null;
    error?: any;
    headers?: object;
    state?: RequestStatus;
};

export function mockApiResponse({
                                    data = {},
                                    status = 200,
                                    error = null,
                                    headers = {},
                                    state = RequestStatus.DONE,
                                }: MockApiResponseOptions = {}) {
    mockUseApi.fetchData.mockResolvedValueOnce({
        ok: () => status === 200,
        data,
        status,
        error,
        headers,
        state
    });
}

export function mockApiResponseSequence(
    responses: MockApiResponseOptions[]
) {
    responses.forEach(
        (mockData) => {
            mockApiResponse(mockData)
        }
    );
}