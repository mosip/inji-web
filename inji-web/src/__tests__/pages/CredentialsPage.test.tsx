import React from 'react';
import {waitFor} from '@testing-library/react';
import {CredentialsPage} from '../../pages/CredentialsPage';
import {mockUseParams, mockUseTranslation, renderWithRouter} from '../../test-utils/mockUtils';
import {mockApiResponse, mockUseApi} from "../../test-utils/setupUseApiMock";
import {RequestStatus} from "../../utils/constants";

mockUseTranslation();
mockUseParams();
//todo : extract the local method to mockUtils, which is added to bypass the routing problems
jest.mock('react-toastify', () => ({
  toast: {
    error: jest.fn(),
  },
}));

jest.mock("../../hooks/useApi.ts", () => {
  return {
    useApi: () => mockUseApi
  };
});

describe('Testing the Layout of CredentialsPage', () => {
  test('Check if the layout is matching with the snapshots', async () => {
    mockApiResponse({
      state: RequestStatus.ERROR,
      error: 'Service Unavailable',
    })
    const { asFragment } = renderWithRouter(<CredentialsPage />);

    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Testing the Functionality of CredentialsPage', () => {
  test('Check if it displays error message if state is ERROR', async () => {
    mockApiResponse({
        state: RequestStatus.ERROR,
        error: 'Service Unavailable',
    })
    renderWithRouter(<CredentialsPage />);

    await waitFor(() => {
      expect(require('react-toastify').toast.error).toHaveBeenCalledWith('The service is currently unavailable now. Please try again later.');
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
});
