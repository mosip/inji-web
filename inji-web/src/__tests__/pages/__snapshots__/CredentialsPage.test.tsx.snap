// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the Layout of CredentialsPage Check if the layout is matching with the snapshots 1`] = `
<DocumentFragment>
  <div
    class="bg-iw-background min-h-screen"
    data-testid="Credentials-Page-Container"
  >
    <div
      class="bg-gradient-to-r from-iw-lightPrimary to-iw-lightSecondary text-iw-title p-4 py-10"
      data-testid="NavBar-Outer-Container"
    >
      <nav
        class=" mx-auto flex flex-col justify-start container items-start sm:flex-row sm:justify-start sm:items-center"
        data-testid="NavBar-Inner-Container"
      >
        <div
          class="flex items-center my-5 sm:my-0"
        >
          <div
            class="cursor-pointer"
          >
            <svg
              data-testid="NavBar-Back-Arrow"
              fill="currentColor"
              height="24"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M244 400 100 256l144-144M120 256h292"
                fill="none"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="48"
              />
            </svg>
          </div>
          <span
            class="text-2xl font-semibold ps-2 pe-4 md:whitespace-nowrap whitespace-wrap"
            data-testid="NavBar-Text"
          />
        </div>
        <div
          class="flex items-center w-full justify-start sm:justify-end my-5 sm:my-0"
          data-testid="NavBar-Search-Container"
        >
          <div
            class="w-full sm:w-96 flex justify-center items-center bg-iw-background shadow-md shadow-iw-shadow"
            data-testid="Search-Issuer-Container"
          >
            <svg
              class="m-5"
              color="var(--iw-color-searchIcon)"
              data-testid="NavBar-Search-Icon"
              fill="currentColor"
              height="22"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="22"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
              />
            </svg>
            <input
              class="py-6 w-11/12 flex text-iw-searchTitle focus:outline-none overflow-ellipsis mr-10"
              data-testid="NavBar-Search-Input"
              placeholder="Search Credential Type"
              type="text"
              value=""
            />
          </div>
        </div>
      </nav>
    </div>
    <div
      class="container mx-auto mt-8 px-10 sm:px-0"
      data-testid="Credential-List-Container"
    >
      <div
        class="flex justify-center items-center"
        data-testid="SpinningLoader-Container"
      >
        <div
          aria-busy="true"
          aria-label="oval-loading"
          class="sc-beySPh gZmPOF"
          data-testid="oval-loading"
          role="progressbar"
        >
          <svg
            data-testid="oval-svg"
            height="80"
            stroke="var(--iw-color-spinningLoaderPrimary)"
            viewBox="-20.5 -20.5 43 43"
            width="80"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fill-rule="evenodd"
            >
              <g
                data-testid="oval-secondary-group"
                stroke-width="3"
                transform="translate(1 1)"
              >
                <circle
                  cx="0"
                  cy="0"
                  r="20"
                  stroke="var(--iw-color-spinningLoaderSecondary)"
                  stroke-opacity=".5"
                  stroke-width="3"
                />
                <path
                  d="M20 0c0-9.94-8.06-20-20-20"
                >
                  <animatetransform
                    attributeName="transform"
                    dur="1s"
                    from="0 0 0"
                    repeatCount="indefinite"
                    to="360 0 0"
                    type="rotate"
                  />
                </path>
              </g>
            </g>
          </svg>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
